elasticsearch/src/http/transport.rs

```diff
@@ -445,7 +469,17 @@ impl Default for Transport {
 /// dynamically at runtime, based upon the response to API calls.
 pub trait ConnectionPool: Debug + dyn_clone::DynClone + Sync + Send {
     /// Gets a reference to the next [Connection]
-    fn next(&self) -> &Connection;
+    fn next(&self) -> Connection;
```

Comment:
Another call out here is the cloning of a current connection. Returning a reference from behind a `MutexGuard` is not possible without returning the whole Guard, making each `ConnectionPool` implementation require locking. I added this as it may make sense to have the current consumer of that current connection own it. Though, I am not sure if there's more overhead in cloning or locking, definitely open to suggestions here. :)

Reply:
On second thought it may be possible to use a generic that implements the `Deref` trait?

Reply:
I would expect the overhead of cloning to be much less than locking.

> On second thought it may be possible to use a generic that implements the Deref trait?

I think this is possible with something like [Rental crate](https://crates.io/crates/rental), but I'm starting to think the cloning approach might be a better option - an Elasticsearch API call might take some time to complete e.g. a complex search query over millions of documents, an update by query with wait_for_completion set to true and a long request timeout, etc. Returning a reference to a `Connection` from a vector of connections inside an `Arc<RwLock<Vec<Connection>>>` would cause a writer to wait to gain exclusive access to update, which could be some time. @swallez, what are your thoughts?

Reply:
+1 `Arc` implementation seems like a great optimization to me!